cluster/addons/dns/kube2sky/kube2sky.go

```diff
@@ -199,41 +216,95 @@ func (ks *kube2sky) addDNSUsingEndpoints(subdomain string, e *kapi.Endpoints) er
 	if err := ks.removeDNS(subdomain); err != nil {
 		return err
 	}
-	return ks.generateRecordsForHeadlessService(subdomain, e, svc)
+	return ks.generateRecordsForHeadlessService(subdomain, e, svc, isNewStyleFormat)
 }
 
 func (ks *kube2sky) handleEndpointAdd(obj interface{}) {
 	if e, ok := obj.(*kapi.Endpoints); ok {
 		name := buildDNSNameString(ks.domain, e.Namespace, e.Name)
-		ks.mutateEtcdOrDie(func() error { return ks.addDNSUsingEndpoints(name, e) })
+		ks.mutateEtcdOrDie(func() error { return ks.addDNSUsingEndpoints(name, e, false) })
 		name = buildDNSNameString(ks.domain, serviceSubdomain, e.Namespace, e.Name)
-		ks.mutateEtcdOrDie(func() error { return ks.addDNSUsingEndpoints(name, e) })
+		ks.mutateEtcdOrDie(func() error { return ks.addDNSUsingEndpoints(name, e, true) })
 	}
 }
 
-func (ks *kube2sky) generateRecordsForPortalService(subdomain string, service *kapi.Service) error {
-	for i := range service.Spec.Ports {
-		b, err := json.Marshal(getSkyMsg(service.Spec.ClusterIP, service.Spec.Ports[i].Port))
+func (ks *kube2sky) generateRecordsForPortalService(subdomain string, service *kapi.Service, isNewStyleFormat bool) error {
+	b, err := json.Marshal(getSkyMsg(service.Spec.ClusterIP, 0))
+	if err != nil {
+		return err
+	}
+	recordValue := string(b)
+	recordKey := subdomain
+	recordLabel := ""
+	if isNewStyleFormat {
+		recordLabel = GetHash(recordValue)
 		if err != nil {
 			return err
 		}
-		glog.V(2).Infof("Setting DNS record: %v -> %q\n", subdomain, string(b))
-		if err := ks.writeSkyRecord(subdomain, string(b)); err != nil {
-			return err
+		recordKey = buildDNSNameString(subdomain, recordLabel)
+	}
+
+	glog.V(2).Infof("Setting DNS record: %v -> %q, with recordKey: %v\n", subdomain, recordValue, recordKey)
+	if err := ks.writeSkyRecord(recordKey, recordValue); err != nil {
+		return err
+	}
+	if !isNewStyleFormat {
+		return nil
+	}
+	// Generate SRV Records
+	for i := range service.Spec.Ports {
+		port := &service.Spec.Ports[i]
+		portSegment := buildPortSegmentString(port.Name, port.Protocol)
+		if portSegment != "" {
+			if err != nil {
+				return err
+			}
+			err = ks.generateSRVRecord(subdomain, portSegment, recordLabel, subdomain, port.Port)
+			if err != nil {
+				return err
+			}
 		}
 	}
 	return nil
 }
 
-func (ks *kube2sky) addDNS(subdomain string, service *kapi.Service) error {
+func buildPortSegmentString(portName string, portProtocol kapi.Protocol) string {
+	if portName == "" {
+		// we don't create a random name
+		return ""
+	}
+	portNameSegment := fmt.Sprintf("_%s", portName)
+
+	if portProtocol == "" {
+		glog.Errorf("Port Protocol not set. port segment string cannot be created.")
+		return ""
+	}
+
+	protocolSegment := fmt.Sprintf("_%s", strings.ToLower(string(portProtocol)))
+	return buildDNSNameString(protocolSegment, portNameSegment)
```

Comment:
minor point but you could just validate input then Sprintf("_%s._%s", portName, strings.ToLower(string(portProtocol))